src/clients/victoriaMetrics/promMetricCounter.ts

```diff
@@ -1,39 +1,99 @@
 import promClient, { Counter } from 'prom-client';
 import { promAgent } from './promAgent';
+import { generateLabelCombinations } from './metricFunctions';
 
-interface CounterConfig {
+export interface CounterConfig {
+  name: string;
+  help: string;
+  labelOptions?: Record<string, Set<string>>;
+}
+
+export interface PromClientCounterConfig {
   name: string;
   help: string;
   labelNames?: string[];
 }
 
 export class PromMetricCounter {
+  private readonly labelOptions: Record<string, Set<string>>;
   private counter: Counter;
-  private labelNames: string[];
 
   constructor(counterConfig: CounterConfig) {
     promAgent.registerMetric(counterConfig.name);
 
-    this.counter = new promClient.Counter(counterConfig);
-    this.labelNames = counterConfig.labelNames;
+    const promClientCounterConfig: PromClientCounterConfig = {
+      name: counterConfig.name,
+      help: counterConfig.help,
+    };
+
+    if (!counterConfig.labelOptions) {
+      this.counter = new promClient.Counter(promClientCounterConfig);
+    } else {
+      if (Object.keys(counterConfig.labelOptions).length === 0) {
+        throw new Error('LabelOptions are declared but undefined.');
+      }
+
+      for (const labelName in counterConfig.labelOptions) {
+        if (counterConfig.labelOptions[labelName].size === 0) {
+          throw new Error(
+            `LabelOptions are undefined for labelName: "${labelName}".`,
+          );
+        }
+      }
+
+      promClientCounterConfig.labelNames = Object.keys(
+        counterConfig.labelOptions,
+      );
+
+      // Initialise counter metric
+      this.labelOptions = counterConfig.labelOptions;
+      this.counter = new promClient.Counter(promClientCounterConfig);
+
+      // Increment each labelCombination by 0 to initiate metric
+      const labelCombinations = generateLabelCombinations(
+        counterConfig.labelOptions,
+      );
+
+      labelCombinations.forEach((combination) => {
+        this.counter.inc(combination, 0);
+      });
+    }
   }
 
   public inc(num: number, labels?: Record<string, string>) {
     if (labels) {
-      this.validateLabels(labels);
+      if (!this.isValidLabelCombination(labels)) {
+        throw new Error(`Invalid labels provided: ${JSON.stringify(labels)}`);
+      }
       this.counter.inc(labels, num);
     } else {
       this.counter.inc(num);
     }
   }
 
-  private validateLabels(labels: Record<string, string>) {
-    const invalidLabels = Object.keys(labels).filter(
-      (label) => !this.labelNames.includes(label),
-    );
+  private isValidLabelCombination(
+    labelCombination: Record<string, string>,
+  ): boolean {
+    // Key size does not match
+    if (
+      Object.keys(labelCombination).length !==
+      Object.keys(this.labelOptions).length
+    ) {
+      return false;
+    }
+
+    for (const labelName in labelCombination) {
+      // label name is not in config
+      if (!this.labelOptions[labelName]) {
```

Comment:
You'll want something like this

`console.log("key" in obj); // true, regardless of the actual value`
From: https://stackoverflow.com/questions/1098040/checking-if-a-key-exists-in-a-javascript-object